### PR TITLE
refactor: avoid clone all table fragments everywhere

### PR DESCRIPTION
**Pull request by yezizp2012**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Part of #11335 . Avoid calling `list_table_fragments` everywhere, because this function will copy all `TableFragments`. The biggest overhead associated with this function is the metrics collection part.

Note that calling `get_fragment_read_guard` incorrectly can easily cause a deadlock, we need to introduce some other functions to replace the original call logic in the future.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/meta/src/dashboard/mod.rs [hunk 1]
```diff
@@ -247,7 +230,6 @@ where
         let api_router = Router::new()
             .route("/clusters/:ty", get(list_clusters::<S>))
             .route("/actors", get(list_actors::<S>))
-            .route("/fragments", get(list_table_fragments::<S>))
```
Comment (BugenZhao) [inline src/meta/src/dashboard/mod.rs:250]: Can we also rename `fragments2` to `fragments`?

File: src/meta/src/stream/source_manager.rs [hunk 2]
```diff
@@ -539,7 +539,12 @@ where
 
         let mut actor_splits = HashMap::new();
         let mut source_fragments = HashMap::new();
-        for table_fragments in fragment_manager.list_table_fragments().await {
+        for table_fragments in fragment_manager
+            .get_fragment_read_guard()
+            .await
+            .table_fragments()
+            .values()
```
Comment (BugenZhao) [inline src/meta/src/stream/source_manager.rs:546]: Just FYI: we may find `LockGuard::map` useful for some helper functions.

File: src/meta/src/stream/stream_graph/fragment.rs [hunk 3]
```diff
@@ -128,6 +128,10 @@ impl BuildingFragment {
                 table.database_id = job.database_id();
                 table.schema_id = job.schema_id();
                 table.fragment_id = fragment_id;
+                #[cfg(not(debug_assertions))]
```
Comment (BugenZhao) [inline src/meta/src/stream/stream_graph/fragment.rs:131]: Where did we do this before? 👀

**Review by BugenZhao — APPROVED** [review]
LGTM

File: src/meta/src/stream/stream_graph/fragment.rs (hunk 3 above)
Comment (yezizp2012) [reply src/meta/src/stream/stream_graph/fragment.rs:131]: We did it before for `mview_definition` in `StreamActor` before. 🥵

**Review by wcy-fdu — APPROVED** [review]
LGTM!

**Pull request merged**